engine/execution/ingestion/stop_control.go

```diff
@@ -0,0 +1,221 @@
+	// For example, blocks at or above stop height has been received, but finalization didn't reach stop height yet.
+	// It can only progress to StopControlPaused
+	StopControlCommenced
+
+	// StopControlPaused means EN has stopped processing blocks. It can happen by reaching set stop height, or
+	// if the node was started with a pause mode.
+	// It is a final state and cannot be changed
+	StopControlPaused
+)
+
+// NewStopControl creates new empty NewStopControl
+func NewStopControl(log zerolog.Logger, paused bool) *StopControl {
+	state := StopControlOff
+	if paused {
+		state = StopControlPaused
+	}
+	log.Debug().Msgf("created StopControl module with paused = %t", paused)
+	return &StopControl{
+		log:   log,
+		state: state,
+	}
+}
+
+// GetState returns current state of StopControl module
+func (s *StopControl) GetState() StopControlState {
+	return s.state
+}
+
+// IsPaused returns true is block execution has been paused
+func (s *StopControl) IsPaused() bool {
+	s.RLock()
+	defer s.RUnlock()
+	return s.state == StopControlPaused
+}
+
+// SetStopHeight sets new stop height and crash mode, and return old values:
+//   - set, whether values were previously set
+//   - height
+//   - crash
+//
+// Returns error is the stopping process has already commenced, new values will be rejected.
+func (s *StopControl) SetStopHeight(height uint64, crash bool) (uint64, bool, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	oldHeight := s.height
+	oldCrash := s.crash
+
+	if s.state == StopControlCommenced {
+		return oldHeight, oldCrash, fmt.Errorf("cannot update stop height, stopping commenced for height %d with crash=%t", oldHeight, oldCrash)
+	}
+
+	if s.state == StopControlPaused {
+		return oldHeight, oldCrash, fmt.Errorf("cannot update stop height, already paused")
+	}
+
+	if s.lastExecutedHeight >= height {
+		return oldHeight, oldCrash, fmt.Errorf("cannot update stop height, given height %d at or below last executed %d", height, s.lastExecutedHeight)
+	}
+
+	s.log.Info().
+		Uint64("height", height).Bool("crash", crash).
+		Uint64("old_height", oldHeight).Bool("old_crash", oldCrash).Msg("new stop height set")
+
+	s.log.Debug().Int8("previous_state", int8(s.state)).Int8("new_state", int8(StopControlSet)).Msg("StopControl state transition")
+	s.state = StopControlSet
+
+	s.height = height
+	s.crash = crash
+	s.stopAfterExecuting = flow.ZeroID
+
+	return oldHeight, oldCrash, nil
+}
+
+// GetStopHeight returns:
+//   - height
+//   - crash
+//
+// Values are undefined if they were not previously set
+func (s *StopControl) GetStopHeight() (uint64, bool) {
+	s.RLock()
+	defer s.RUnlock()
+
+	return s.height, s.crash
+}
+
+// BlockProcessable should be called when new block is processable.
+// It returns boolean indicating if the block should be processed.
+func (s *StopControl) BlockProcessable(b *flow.Header) bool {
+
+	s.Lock()
+	defer s.Unlock()
+
+	if s.state == StopControlOff {
+		return true
```

Comment:
I think there is a unhappy path missed here, please refer to my comment for `highestProcessableHeight` and `NotifyBlockProcessable`. 

The idea of `highestProcessableHeight` is the same as @AlexHentschel 's idea for using `latestScheduledForExecution`. Let me call it `highestProcessableHeight` for now.

We need to always update the highest processable height unless it's paused, that's why I handle the Paused state first.
Also note, it's "highest" or "latest". The latest schedule could be a lower height than the highest. 

So there are a few test cases we need to add and pass:

1. if stop control has no stop height, then called BlockProcessable(20), and then set stop height 20, it should fail, because stop height 20 means the last block to execute is 19, but a 20 might have been processed. 
2. if stop control has no stop height, then called BlockProcessable(20), and then set stop height 21, it should succeed, because stop height 21 means the last block to execute is 20. 
3. if stop control has no stop height, then called BlockProcessable(20), and then called BlockProcessable(19), then set stop height 20 should fail, because we use highest processable height to guard, rather than latest.

Reply:
Scheduled for execution is a bad approach in my mind.
Imagine EN which is far behind. Behind it has not executed blocks for the very same network it is a part of, for a significant number of blocks. 
It can schedule a large number of blocks for execution, sometime hours or days' worth of blocks. This would mean it's impossible to set a stop height to anything between current executed block and the actual latest block.

Additionally, I see no reason why we would prevent from stopping even one block from what's possible to properly stop, which is highest currently being executed+1.
I think practical solution would be to track highest block at the start of execution.